torchqdynamics/mesolve.py

```diff
@@ -1,63 +1,103 @@
-    return torch.view_as_complex(ys)
+    H, jump_ops, rho0, tsave, solver=Rouchon(), sensitivity='autograd',
+    model_params=None
+):
+    # Define the QSolver
+    if isinstance(solver, Rouchon):
+        if solver.order == 1:
+            qsolver = MERouchon1(H, jump_ops, solver)
+        elif solver.order == 2:
+            qsolver = MERouchon2(H, jump_ops, solver)
+    else:
+        raise NotImplementedError
+
+    # Compute the result
+    return odeint(
+        qsolver, rho0, tsave, sensitivity=sensitivity, model_params=model_params
+    )
+
+
+# --------------------------------------------------------------------------------------
+#     ME QSolver classes
+# --------------------------------------------------------------------------------------
+
+
+class QSolver:
+    def __init__(self):
+        pass
+
+
+class MERouchon(QSolver):
+    def __init__(self, H, jump_ops, solver_options):
+        self.H = H
+        self.jump_ops = jump_ops
+        self.jumpdag_ops = jump_ops.adjoint()
+        self.sum_nojump = (self.jumpdag_ops @ self.jump_ops).sum(dim=0)
+        self.I = torch.eye(H(0).shape[-1]).to(H(0))
+        self.options = solver_options
+
+
+class MERouchon1(MERouchon):
+    def forward(self, t, dt, rho):
+        """Compute rho(t+dt) using a Rouchon method of order 1"""
+        # Non-hermitian Hamiltonian at time t
+        H_nh = self.H(t) - 0.5j * self.sum_nojump
+
+        # Build time-dependent Kraus operators
+        M0 = self.I - 1j * dt * H_nh
+        M1s = self.jump_ops
+
+        # Compute rho(t+dt)
+        drho = M0 @ rho @ M0.adjoint(
+        ) + dt * (self.jump_ops @ rho.unsqueeze(0) @ self.jumpdag_ops).sum(dim=0)
+
+        return drho
+
+    def forward_adjoint(self, t, dt, phi):
+        raise NotImplementedError
+
+
+class MERouchon2(MERouchon):
+    def forward(self, t, dt, rho):
+        """Compute rho(t+dt) using a Rouchon method of order 2"""
+        # Non-hermitian Hamiltonian at time t
+        H_nh = self.H(t) - 0.5j * self.sum_nojump
+
+        # Build time-dependent Kraus operators
+        M0 = self.I - 1j * dt * H_nh - 0.5 * dt**2 * H_nh**2
+        M1s = 0.5 * (self.Ls @ M0 + M0 @ self.Ls)
+
+        # Compute rho(t+dt)
+        drho_ = dt * torch.sum(M1s @ rho.unsqueeze(0) @ M1s.adjoint(), dim=0)
+        drho += 0.5 * dt * torch.sum(M1s @ drho_.unsqueeze(0) @ M1s.adjoint(), dim=0)
+        drho += drho_ + M0 @ rho @ M0.adjoint()
+
+        return drho
+
+    def forward_adjoint(self, t, dt, phi):
+        raise NotImplementedError
+
+
+# --------------------------------------------------------------------------------------
+#     ME Solver Options
+# --------------------------------------------------------------------------------------
+
+# See the PR by @abocquet at https://github.com/PierreGuilmin/torchqdynamics/pull/10
+
+
+@dataclass
+class Rouchon:
+    dt: float = 1e-2
+    order: int = 1
+    stepclass: str = 'fixed'
```

Comment:
Using strings for enumeration is a pattern I'm ill at ease with because we don't know what are the accepted values. Here I'm guessing `stepclass` could be `fixed` or "fixed but variable in time"?

Reply:
I was thinking that `stepclass` should be either `fixed` or `adaptive` here. It's just an information so that we know whether we should dispatch to the `fixed_odeint` or `adaptive_odeint` later on. In your solver options PR you have the two separate classes for fixed and adaptive, so we could use just the parent class name instead of a string class argument?

For time-dependence, I think it should be implemented directly in the QSolver. Something like `MERouchon1` and `MERouchon1_TD` maybe? Or it could be the same class, but just pre-compute the Kraus operators in the `Qsolver.__init__`?

Reply:
+1 on the separate classes if needed, but I thought the Rouchon method could only support fixed time steps (or time dependent but fixed in advanced).

If all fixed timestep method support both a simple dt and a time depedent timestep then we could simply say that the `step` argument accepts a float or a list of floats?

Reply:
No that's not what I meant the `stepclass` argument for. I meant it to be used here: https://github.com/PierreGuilmin/torchqdynamics/blob/f00ac9e980515c434891cc2c394e81eecbbd2198/torchqdynamics/odeint.py#L31-L34
just to dispatch to the related fixed or adaptive odeint solver. It's unrelated to time-dependence or varying dt.

Reply:
Yes, agreed. Are there methods that can be both used with a fixed timestep or an adaptative one?
For instance I think that Rouchon method is fixed only and DOPRI6 is adaptative only?

Reply:
No I think it's always one or the other.

Reply:
Then can't be `stepclass` inferred from the schema?

Reply:
Yes but I didn't know how to, and I also wanted a simple implementation for this PR. But it should be implemented properly from your PR instead.

Reply:
Ok good to know, I just wanted to make sure I didn't miss something.
Let's do this after we merge https://github.com/PierreGuilmin/torchqdynamics/pull/10